File: labs/D80_Builder/python/complete/housebuilder.py
from floor import Floor
from foundation import Foundation
from roof import Roof
from wall import Wall
from house import House
from typing import List


class HouseBuilder():
    foundation: Foundation = None
    floors: List[Floor] = []
    walls: List[Wall] = []
    roof: Roof = None

    def __init__(self):
        pass

    def is_buildable(self)->bool:
        return (self.foundation is not None) and \
               len(self.walls) > 3 and \
               self.roof is not None and \
               len(self.floors) > 0


    def to_house(self)->House:
        return House(self.foundation,self.floors,self.walls,self.roof)

    def withInteriorWalls(self, quantity:int):
        for i in range(0, quantity):
            newWall = Wall('Drywall', False, False)
            self.walls.append(newWall)
        return self

    def withExteriorWalls(self, material:str, quantity:int = 4):
        for i in range(0, quantity):
            newWall = Wall(material, True, True)
            self.walls.append(newWall)
        return self

    def withWoodFloor(self):
        self.floors = [
            Floor('Wood')
        ]
        return self

    def withConcreteFoundation(self):
        self.foundation=Foundation('Concrete')
        return self

    def withRegularRoof(self):
        self.roof = Roof('Shingles')
        return self

File: labs/D80_Builder/java/complete/app/src/main/java/BuilderComplete/HouseBuilder.java
package BuilderComplete;

import java.util.ArrayList;
import java.util.List;

public class HouseBuilder {
  private List<Wall> walls = new ArrayList<Wall>();
  private List<Floor> floors = new ArrayList<Floor>();
  private Roof roof;
  private Foundation foundation;
  
  public boolean isBuildable(){
    if(roof==null){
      return false;
    }

    if(foundation==null){
      return false;
    }

    return walls.size()>3 && floors.size()>0;
  }

  public HouseBuilder withInteriorWalls(int quantity){
    for(int i=0;i<quantity;i++){
      this.walls.add(new Wall("Drywall",false,false));
    }
    return this;
  }

  public HouseBuilder withExteriorWalls(int quantity, String material){
    for(int i=0;i<quantity;i++){
      this.walls.add(new Wall(material,true,true));
    }
    return this;
  }

  public HouseBuilder withExteriorWalls(int quantity){
    return this.withExteriorWalls(quantity,"Wood");
  }


  public HouseBuilder withWoodFloor(){
    floors.add(new Floor("Wood"));
    return this;
  }
  
  public HouseBuilder withConcreteFoundation(){
    foundation = new Foundation("Concrete");
    return this;
  }
  
  public HouseBuilder withRegularRoof(){
    roof = new Roof("Shingles");
    return this;
  }


  public House toHouse() throws Exception{
    if(!this.isBuildable()){
      throw new Exception("Bogus state!");
    }
    return new House(roof, foundation, walls.toArray(new Wall[0]), floors.toArray(new Floor[0]));
  }

}
